File: src/common/py/api/project_commands.py
import dataclasses
import typing

from ..core.messaging import (
    Command,
    CommandReply,
    Message,
)
from ..core.messaging.composers import (
    MessageBuilder,
    CommandComposer,
    CommandReplyComposer,
)
from ..data.entities import Project, ProjectID, ProjectFeatureStore, ProjectFeatureID


@Message.define("command/project/list")
class ListProjectsCommand(Command):
    """
    Command to fetch all projects of the current user.

    Notes:
        Requires a ``ListProjectsReply`` reply.
    """

    @staticmethod
    def build(
        message_builder: MessageBuilder, *, chain: Message | None = None
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(ListProjectsCommand, chain)


@Message.define("command/project/list/reply")
class ListProjectsReply(CommandReply):
    """
    Reply to ``ListProjectsCommand``.

    Args:
        projects: List of all projects.
    """

    projects: typing.List[Project] = dataclasses.field(default_factory=list)

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: ListProjectsCommand,
        *,
        projects: typing.List[Project],
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            ListProjectsReply, cmd, success, message, projects=projects
        )


@Message.define("command/project/create")
class CreateProjectCommand(Command):
    """
    Command to create a new project.

    Args:
        title: The title of the project.
        description: An optional project description.
        features_selection: List of enabled user-selectable features.

    Notes:
        Requires a ``CreateProjectReply`` reply.
    """

    title: str
    description: str

    features_selection: typing.List[ProjectFeatureID] = dataclasses.field(
        default_factory=list
    )

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        title: str,
        description: str,
        features_selection: typing.List[ProjectFeatureID],
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            CreateProjectCommand,
            chain,
            title=title,
            description=description,
            features_selection=features_selection,
        )


@Message.define("command/project/create/reply")
class CreateProjectReply(CommandReply):
    """
    Reply to ``CreateProjectCommand``.

    Args:
        project_id: The ID of the created project.
    """

    project_id: ProjectID

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: CreateProjectCommand,
        *,
        project_id: ProjectID,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            CreateProjectReply, cmd, success, message, project_id=project_id
        )


@Message.define("command/project/update")
class UpdateProjectCommand(Command):
    """
    Command to update an existing project.
    Note that the project features are updated using a separate ``UpdateProjectFeaturesCommand`` message.

    Args:
        project_id: The ID of the project to update.
        title: The title of the project.
        description: An optional project description.
        features_selection: List of enabled user-selectable features.

    Notes:
        Requires an ``UpdateProjectReply`` reply.
    """

    project_id: ProjectID

    title: str
    description: str

    features_selection: typing.List[ProjectFeatureID] = dataclasses.field(
        default_factory=list
    )

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        project_id: ProjectID,
        title: str,
        description: str,
        features_selection: typing.List[ProjectFeatureID],
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            UpdateProjectCommand,
            chain,
            project_id=project_id,
            title=title,
            description=description,
            features_selection=features_selection,
        )


@Message.define("command/project/update/reply")
class UpdateProjectReply(CommandReply):
    """
    Reply to ``UpdateProjectCommand``.

    Args:
        project_id: The ID of the updated project.
    """

    project_id: ProjectID

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: UpdateProjectCommand,
        *,
        project_id: ProjectID,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            UpdateProjectReply, cmd, success, message, project_id=project_id
        )


@Message.define("command/project/delete")
class DeleteProjectCommand(Command):
    """
    Command to delete a project of the current user.

    Args:
        project_id: The ID of the project to delete.

    Notes:
        Requires a ``DeleteProjectReply`` reply.
    """

    project_id: ProjectID

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        *,
        project_id: ProjectID,
        chain: Message | None = None,
    ) -> CommandComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command(
            DeleteProjectCommand, chain, project_id=project_id
        )


@Message.define("command/project/delete/reply")
class DeleteProjectReply(CommandReply):
    """
    Reply to ``DeleteProjectCommand``.

    Args:
        project_id: The ID of the deleted project.
    """

    project_id: ProjectID

    @staticmethod
    def build(
        message_builder: MessageBuilder,
        cmd: DeleteProjectCommand,
        *,
        project_id: ProjectID,
        success: bool = True,
        message: str = "",
    ) -> CommandReplyComposer:
        """
        Helper function to easily build this message.
        """
        return message_builder.build_command_reply(
            DeleteProjectReply, cmd, success, message, project_id=project_id
        )

File: src/common/web/api/ProjectCommands.ts
import { Type } from "class-transformer";

import { Command } from "../core/messaging/Command";
import { CommandReply } from "../core/messaging/CommandReply";
import { CommandComposer } from "../core/messaging/composers/CommandComposer";
import { CommandReplyComposer } from "../core/messaging/composers/CommandReplyComposer";
import { MessageBuilder } from "../core/messaging/composers/MessageBuilder";
import { Message } from "../core/messaging/Message";
import { type ProjectFeatureID, type ProjectID } from "../data/entities/EntityTypes";
import { Project } from "../data/entities/Project";

/**
 * Command to fetch all projects of the current user. Requires a ``ListProjectsReply`` reply.
 */
@Message.define("command/project/list")
export class ListProjectsCommand extends Command {
    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, chain: Message | null = null): CommandComposer<ListProjectsCommand> {
        return messageBuilder.buildCommand(ListProjectsCommand, {}, chain);
    }
}

/**
 * Reply to ``ListProjectsCommand``.
 *
 * @param projects - The projects list.
 */
@Message.define("command/project/list/reply")
export class ListProjectsReply extends CommandReply {
    // @ts-ignore
    @Type(() => Project)
    public readonly projects: Project[] = [];

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: ListProjectsCommand,
        projects: Project[],
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<ListProjectsReply> {
        return messageBuilder.buildCommandReply(ListProjectsReply, cmd, success, message, { projects: projects });
    }
}

/**
 * Command to create a project. Requires a ``CreateProjectReply`` reply.
 *
 * @param title - The title of the project.
 * @param description - An optional project description.
 * @param features_selection - List of enabled user-selectable features.
 */
@Message.define("command/project/create")
export class CreateProjectCommand extends Command {
    public readonly title: string = "";
    public readonly description: string = "";

    // @ts-ignore
    @Type(() => String)
    public readonly features_selection: ProjectFeatureID[] = [];

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        title: string,
        description: string,
        features_selection: ProjectFeatureID[],
        chain: Message | null = null
    ): CommandComposer<CreateProjectCommand> {
        return messageBuilder.buildCommand(CreateProjectCommand, { title: title, description: description, features_selection: features_selection }, chain);
    }
}

/**
 * Reply to ``CreateProjectCommand``.
 *
 * @param project_id: The ID of the created project.
 */
@Message.define("command/project/create/reply")
export class CreateProjectReply extends CommandReply {
    public readonly project_id: ProjectID = 0;

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: CreateProjectCommand,
        project_id: ProjectID,
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<CreateProjectReply> {
        return messageBuilder.buildCommandReply(CreateProjectReply, cmd, success, message, { project_id: project_id });
    }
}

/**
 * Command to update a project. Requires an ``UpdateProjectReply`` reply.
 * Note that the project features are updated using a separate ``UpdateProjectFeaturesCommand`` message.
 *
 * @param project_id - The ID of the project to update.
 * @param title - The title of the project.
 * @param description - An optional project description.
 * @param features_selection - List of enabled user-selectable features.
 */
@Message.define("command/project/update")
export class UpdateProjectCommand extends Command {
    public readonly project_id: ProjectID = 0;

    public readonly title: string = "";
    public readonly description: string = "";

    // @ts-ignore
    @Type(() => String)
    public readonly features_selection: ProjectFeatureID[] = [];

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        project_id: ProjectID,
        title: string,
        description: string,
        features_selection: ProjectFeatureID[],
        chain: Message | null = null
    ): CommandComposer<UpdateProjectCommand> {
        return messageBuilder.buildCommand(UpdateProjectCommand, {
            project_id: project_id, title: title, description: description, features_selection: features_selection
        }, chain);
    }
}

/**
 * Reply to ``UpdateProjectCommand``.
 *
 * @param project_id - The ID of the updated project.
 */
@Message.define("command/project/update/reply")
export class UpdateProjectReply extends CommandReply {
    public readonly project_id: ProjectID = 0;

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: UpdateProjectCommand,
        project_id: ProjectID,
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<UpdateProjectReply> {
        return messageBuilder.buildCommandReply(UpdateProjectReply, cmd, success, message, { project_id: project_id });
    }
}

/**
 * Command to delete a project of the current user. Requires a ``DeleteProjectReply`` reply.
 *
 * @param project_id - The ID of the project to delete.
 */
@Message.define("command/project/delete")
export class DeleteProjectCommand extends Command {
    public readonly project_id: ProjectID = 0;

    /**
     * Helper function to easily build this message.
     */
    public static build(messageBuilder: MessageBuilder, projectID: ProjectID, chain: Message | null = null): CommandComposer<DeleteProjectCommand> {
        return messageBuilder.buildCommand(DeleteProjectCommand, { project_id: projectID }, chain);
    }
}

/**
 * Reply to ``DeleteProjectCommand``.
 *
 * @param project_id - The ID of the deleted project.
 */
@Message.define("command/project/delete/reply")
export class DeleteProjectReply extends CommandReply {
    public readonly project_id: ProjectID = 0;

    /**
     * Helper function to easily build this message.
     */
    public static build(
        messageBuilder: MessageBuilder,
        cmd: DeleteProjectCommand,
        success: boolean = true,
        message: string = ""
    ): CommandReplyComposer<DeleteProjectReply> {
        return messageBuilder.buildCommandReply(DeleteProjectReply, cmd, success, message, { project_id: cmd.project_id });
    }
}
